### ISSUE TITLE
Write a "how it works" documentation

### ISSUE THREAD

**Issue by vincentfretin** [networked-aframe/networked-aframe#409]
Write an overview of how it works “internally”, regarding data-flow, serialization, send interval, if it's send at all (ownership), deserialization on the receiving end. A diagram would be nice.
What do you folks want to know?

**Comment by 2xAA** [comment]
Information about the connection/handshake would be nice, so people can easily write adapters :)

**Comment by vincentfretin** [comment]
The connection to a signaling server and connecting to other participants depends on the WebRTC server tech you use, but I know what you mean. To create an adapter you need to implement a set of methods, you have the list on the interface [NoOpAdapter](https://github.com/networked-aframe/networked-aframe/blob/master/src/adapters/NoOpAdapter.js)
You can look at the [different adapters in this repo](https://github.com/networked-aframe/networked-aframe/tree/master/src/adapters) to have a sense of what is needed in those methods.

**Comment by 2xAA** [comment]
True, I've just created a slimline WebSocket server as I was finding easyrtc too opaque as an easy "get to grips" solution.

What helped me to understand everything in the end was putting a console.log in each of the existing wseasyrtc adapter's methods to see the exact flow and expected callbacks.

For example, I didn't really understand when a client would be sent an individual message vs a broadcast, and if they should be handled separately (the answer is no, they should both be fed to the messageListener callback in setDataChannelListeners).

Another issue I had was the setRoom function, it seems like the connection and setRoom are independent, but with wseasyrtc at least, the adapter expects the room to be joined on connection, before it returns the clientid - that's not obvious unless you debug quite closely.

I hope this is useful feedback!

**Comment by tomfelder94** [comment]
I also find it quite hard to get a good understanding of the whole flow, when is which method called from the adapters. Also when is information broadcasted to the channel, i.e. in a client-server kind of way, and what information is sent p2p. I will also try to understand better adding a console.log to all the different methods in the adapter. A visual overview/ diagram as you proposed would be much appreciated, whenever you find time.

**Comment by vincentfretin** [comment]
Alright I'll focus first on the transfer logic for the entities data, I won't go in details on the signaling part in this post.
All the transferred data are the entities, mostly all is handled in the [networked](https://github.com/networked-aframe/networked-aframe/blob/master/src/components/networked.js) component, it uses the special `u` (update), `um` (update multiple) an `r` (remove) messages.
https://github.com/networked-aframe/networked-aframe/blob/32996226c8b8f4e21be154090844d24e2d182696/src/NetworkConnection.js#L2

When you connect, each networked component will take care of sending the entity to all participants if you're the owner of the entity. When you create an entity, at first you're the creator and owner of it, creator won't change, owner can change during the lifetime of the entity.
Here when I say 'sending the entity' I mean a json containing the full data or a subset of the data of the components of the networked entity, potentially components of its children, based on the networked schema associated with the networked entity.

https://github.com/networked-aframe/networked-aframe/blob/32996226c8b8f4e21be154090844d24e2d182696/src/components/networked.js#L277 is called via the `onConnected` callback registered for the the `connected` event that is triggered here
https://github.com/networked-aframe/networked-aframe/blob/32996226c8b8f4e21be154090844d24e2d182696/src/NetworkConnection.js#L70-L76

This call `syncAll(undefined, true)` meaning send to all participants the owned entities and with data marked with `isFirstSync: true`, will call 
https://github.com/networked-aframe/networked-aframe/blob/32996226c8b8f4e21be154090844d24e2d182696/src/components/networked.js#L342

When a user join, all the participants are notified, if those other participants own some entities, they will send their owned entities to the joined user with 
https://github.com/networked-aframe/networked-aframe/blob/32996226c8b8f4e21be154090844d24e2d182696/src/components/networked.js#L340
This is actually when the user opens a connection to the new user that we do the firstSync to this user only
https://github.com/networked-aframe/networked-aframe/blob/32996226c8b8f4e21be154090844d24e2d182696/src/NetworkConnection.js#L130-L137
that calls 
https://github.com/networked-aframe/networked-aframe/blob/32996226c8b8f4e21be154090844d24e2d182696/src/NetworkEntities.js#L141-L147

Later, if you take ownership of an entity (with the `takeOwnership` api), it calls `syncAll()` to broadcast to all participants the change, it will use 
https://github.com/networked-aframe/networked-aframe/blob/32996226c8b8f4e21be154090844d24e2d182696/src/components/networked.js#L342
but this time with `isFirstSync: false`on the data.

When an entity is removed by the owner, it broadcasts a remove message to all participants:
https://github.com/networked-aframe/networked-aframe/blob/32996226c8b8f4e21be154090844d24e2d182696/src/components/networked.js#L581

All the above use the reliable transport. Depending of the adapter it can be the websocket or a WebRTC datachannel configured with `{ ordered: true }`

When you do some changes to a owned entity (like position and rotation or other components defined in the networked schema), it uses
https://github.com/networked-aframe/networked-aframe/blob/32996226c8b8f4e21be154090844d24e2d182696/src/components/networked.js#L104
to broadcast to all participants using the unreliable transport, see no Guaranteed in the name of this api.
Depending of the adapter it can be websocket or WebRTC datachannel configured with `{ ordered: false, maxRetransmits: 0 }`. 

When you receive a notification that a user disconnected, all non persistent entities he created even if owned by someone else are removed. If there are persistent entities that the user created, each participant will try to get ownership of those persistent entities, the participant the quicker wins.
This is this part in the code:
https://github.com/networked-aframe/networked-aframe/blob/32996226c8b8f4e21be154090844d24e2d182696/src/NetworkConnection.js#L139-L146
that calls 
https://github.com/networked-aframe/networked-aframe/blob/32996226c8b8f4e21be154090844d24e2d182696/src/NetworkEntities.js#L155-L172

Feel free to contribute a diagram from the above explanation. :) You can start one with https://excalidraw.com and share it here if you want.

**Comment by vincentfretin** [comment]
Indeed with the `broadcastData` and `broadcastDataGuaranteed` API to broadcast data to everyone, depending of the adapter, can be sending to each participant peer to peer (mesh topology) or once to a server that broadcasts to all participants (SFU topology).

Only EasyRTCAdapter is peer-to-peer (audio/video/data with WebRTC), mesh topology.
WsEasyRtcAdapter is data only via WebSocket, so client->server->clients.
Janus adapter is a SFU, uploading only once audio/video with WebRTC to a server, data via WebRTC or WebSocket (you can choose). With janus adapter you can have 20-25 users with audio in a room without issue.

**Comment by tomfelder94** [comment]
That makes sense, thanks. I will implement an Ably adapter and see if I find time to create a diagram along the way. That was all very helpful!

**Comment by tomfelder94** [comment]
I created an AblyAdapter today which seems to work but I still have a few questions:

- I am connecting to ably in the adapters `connect()` method. The `broadcastDataGuaranteed(dataType, data)` method is called immediately after, ideally we could await a promise that is resolved once connection is established. I looked at how it is done in the socketio adapter, there you are simply checking if the socket already exists or not. Should I do it this way according to best practice principles? 

https://github.com/networked-aframe/networked-aframe/blob/32996226c8b8f4e21be154090844d24e2d182696/src/adapters/naf-socketio-adapter.js#L170-L183
- What are the `shouldStartConnectionTo(clientId)` and `getConnectStatus(clientId)` methods exactly here for? It seems like they are not needed for the AblyAdapter.
- Do I understand correctly, that the three "stream" methods (see below) are only used if we stream media using WebRTC?
```
    startStreamConnection(clientId) 
    closeStreamConnection(clientId) 
    getMediaStream(clientId) 
```

**Comment by vincentfretin** [comment]
For easyrtc, the check is done in the open-easyrtc library
https://github.com/open-easyrtc/open-easyrtc/blob/221f6439a083aa6e47b0e508d155723b60bdec4e/api/easyrtc_int.js#L3251-L3253

There is a similar check in janus adapter for publisher WebRTC connection
https://github.com/networked-aframe/naf-janus-adapter/blob/fed7925cd1f4f771980f63a845edcfd606ae3544/src/index.js#L1016-L1019

In theory you should never see those warnings, If you see them, there is probably something wrong.
Now that you mention it, I have sometimes the warning with the janus adapter but never looked why, although the code in the adapter seems to do the right thing.

As we saw above, the initial `broadcastDataGuaranteed` call is triggered from the `this.syncAll(undefined, true)`  call from the `connectSuccess` callback set via
https://github.com/networked-aframe/networked-aframe/blob/32996226c8b8f4e21be154090844d24e2d182696/src/NetworkConnection.js#L46-L49
In the case of the janus adapter, the socket is opened, the publisher WebRTC is created and then the connectSuccess callback is called. So you definitely should do something like that if your reliable transport is via WebRTC datachannel
https://github.com/networked-aframe/naf-janus-adapter/blob/fed7925cd1f4f771980f63a845edcfd606ae3544/src/index.js#L246-L252
The broadcast for the first sync via the WebRTC datachannel initially is probably useless for janus sfu, because nobody is connected to us yet, so the real first sync will happen when we're connecting to each participant and when the `onOccupantConnected` callback is called
https://github.com/networked-aframe/naf-janus-adapter/blob/fed7925cd1f4f771980f63a845edcfd606ae3544/src/index.js#L344

For the easyrtc adapter, the `connectSucess` callback is called by the library after the websocket is opened
https://github.com/open-easyrtc/open-easyrtc/blob/221f6439a083aa6e47b0e508d155723b60bdec4e/api/easyrtc_int.js#L6085
the broadcast is done peer to peer:
https://github.com/networked-aframe/networked-aframe/blob/7d7b1abd5e25af05ef337d315d26e94359080893/src/adapters/EasyRtcAdapter.js#L166-L175
It will actually send the first sync via websocket if the WebRTC datachannel is not yet ready.
https://github.com/open-easyrtc/open-easyrtc/blob/221f6439a083aa6e47b0e508d155723b60bdec4e/api/easyrtc_int.js#L3273-L3280

**Comment by vincentfretin** [comment]
For the `shouldStartConnectionTo(clientId)`, `startStreamConnection(clientId)`, `closeStreamConnection(clientId)` api
some adapter may use those methods to implement the connection and disconnection logic to participants, this is based on the `occupantsReceived` callback:
https://github.com/networked-aframe/networked-aframe/blob/32996226c8b8f4e21be154090844d24e2d182696/src/NetworkConnection.js#L82-L108
The easyrtc adapter uses those methods:
https://github.com/networked-aframe/networked-aframe/blob/7d7b1abd5e25af05ef337d315d26e94359080893/src/adapters/EasyRtcAdapter.js#L127-L150

janus adapter doesn't use those methods and do the logic internally:
https://github.com/networked-aframe/naf-janus-adapter/blob/fed7925cd1f4f771980f63a845edcfd606ae3544/src/index.js#L819-L825

Those methods are used to subscribe to a participant for naf updates, audio, video.
In the case of easyrtc a WebRTC connection to a given participant is created to receive the naf updates via datachannel, audio and video tracks.
For janus sfu, it's also creating a WebRTC connection for each participant, although going really through the server, not peer to peer.
For another sfu like LiveKit or mediasoup where it's using only two WebRTC connections (one to publish, one to receive), it can be used to subscribe to a participant datachannel, audio track and video track.

`getMediaStream(clientId, streamName)` api is used for audio and video, it's used in the the `networked-audio-source` and `networked-video-source` components.
